src/daemon/daemon.cpp

```diff
@@ -1132,6 +1153,41 @@ bool is_ipv4_valid(const std::string& ipv4)
     return true;
 }
 
+struct SnapshotPick
+{
+    std::unordered_set<std::string> pick;
+    bool all_or_none;
+};
+using InstanceSnapshotPairs = google::protobuf::RepeatedPtrField<mp::InstanceSnapshotPair>;
+using InstanceSnapshotsMap = std::unordered_map<std::string, SnapshotPick>;
+InstanceSnapshotsMap map_snapshots_to_instances(const InstanceSnapshotPairs& instances_snapshots)
+{
+    InstanceSnapshotsMap instance_snapshots_map;
+
+    for (const auto& it : instances_snapshots)
+    {
+        const auto& instance = it.instance_name();
+        auto& snapshot_pick = instance_snapshots_map[instance];
+
+        if (!it.has_snapshot_name())
+            snapshot_pick.all_or_none = true;
+        else
+            snapshot_pick.pick.insert(it.snapshot_name());
+    }
+
+    return instance_snapshots_map;
+}
+
+void verify_snapshot_picks(const InstanceSelectionReport& report,
+                           const std::unordered_map<std::string, SnapshotPick>& snapshot_picks)
+{
+    for (const auto& selection : {report.deleted_selection, report.operative_selection})
+        for (const auto& vm_it : selection)
+            if (auto pick_it = snapshot_picks.find(vm_it->first); pick_it != snapshot_picks.end())
+                for (const auto& snapshot_name : pick_it->second.pick)
```

Comment:
Well, nothing really comes to mind, so let's just keep.  I will say that although it is a bit hard to read and follow it, the code is tight and is probably as optimized as it can be.